### PR TITLE
Changelog meta entry

### PR DESCRIPTION
**Pull request by 7c6f434c**
###### Motivation for this change

There are various calls to pay attention to changelog. For easier location of the changelog and for better tooling support, we should allow putting a link to the changelog into `meta`.

This only changes the documentation and the `meta` correctness check.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- N/A Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [X] Assured whether relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Review by grahamc — APPROVED** [review]
Nice idea.

Can you add a changelog entry to a single package in this PR as well, to ensure ofborg gets along with it?

**Comment by 7c6f434c** [comment]
Given that the example in the documentation already works, added it to `hello` itself, too.

**Comment by c0bw3b** [comment]
There was a related discussion in #50483

**Comment by 7c6f434c** [comment]
Hm, a good question: should we allow null for lack of changelog, or should it be empty list, or just a comment?

File: pkgs/applications/misc/hello/default.nix
Comment (ryantm) [inline pkgs/applications/misc/hello/default.nix]:
```suggestion
    changelog = "https://git.savannah.gnu.org/cgit/hello.git/plain/NEWS?h=v${version}";
```

File: doc/meta.xml
Comment (ryantm) [inline doc/meta.xml]:
```suggestion
      <literal>"https://git.savannah.gnu.org/cgit/hello.git/plain/NEWS?h=v${version}"</literal>
```

I think I prefer the plain text changelog link. Any reason not to use that?

**Review by ryantm — CHANGES_REQUESTED** [review]
nixpkgs-update can use this to provide a link to the changelog. I'm not sure how much this will help to get the text of the changes into the report. It would depend on how easy it is to parse changelogs, or if it is good enough to include say the first 100 lines of the changelog file.

**Comment by 7c6f434c** [comment]
1. Thanks for the plaintext link suggestion — I found something that looks mostly OK and stopped there, plaintext is indeed better.

2. I think a link to the changelog will be quite useful. 100 first lines sounds like it could be useful more often than not, but it might be suboptimal in case only HTML page (with header etc.) is available.

**Comment by c0bw3b** [comment]
> 
> 
> Hm, a good question: should we allow null for lack of changelog, or should it be empty list, or just a comment?

I guess an empty string (or empty list) would be expected when no `changelog` is given.
Similarly to not setting any homepage or any description.

**Comment by 7c6f434c** [comment]
@c0bw3b I do not think there has been a push to set `meta.homepage` for a maximum possible number of packages; it looks like some people could try such an effort for `meta.changelog`, so I wonder how to distinguish «there is no such thing» and «we haven't looked for it yet» in a way that is friendly to tooling.

**Review by ryantm — APPROVED** [review]

**Comment by ryantm** [comment]
I approve the changes; of course, the commit messages should get fixed up before merging.

**Comment by 7c6f434c** [comment]
My current plan is indeed to squash-merge.

**Comment by grahamc** [comment]
Ofborg could and should send inline comments suggesting filling out maintainer, homepage, change log, license fields.

**Comment by 7c6f434c** [comment]
Hopefully license always exists (otherwise it is unfree by definition), but homepage can fail to exist and maybe there should be a documented way to make a claim it doesn't exist. I think empty list is better than empty string; no opinion on allowing `null` instead of `[]`.

**Review by matthewbauer — APPROVED** [review]

**Comment by 7c6f434c** [comment]
OK, the current plan us to assume that as `homepage` and `changelog` can be lists of alternatives, `[]` means that we looked for a page but found nothing. Right now there is no need to document it; once some tooling starts demanding these attributes, we will document it in a place that makes the most sense in the new context.

(The current PR will be squash-merged as-is)

**Comment by timokau** [comment]
I just tried to add a changelog to `ntl`. The changelog is available [here](https://www.shoup.net/ntl/doc/tour-changes.html). However linking to that page has two disadvantages:

- its not possible to directly link to the current version
- there is no guarantee that the page will be available in the future

The changelog is also available in `$src/doc/tour-changes.html` or `$out/share/doc/NTL/tour-changes.html`. It would be much nicer to link to one of those instead. We already mirror the source, so availability guarantees are much better. Also the relevant version will always be at the top.

However as far as I understand pointing to `$out` is problematic in a meta entry, since it requires a build in order to evaluate. Pointing to `src` is also problematic, even if a bit less so.

Opinions?

**Comment by 7c6f434c** [comment]
I think that `meta.changelog` is either for humans or for tools processing fresh changes.

Therefore I would set the changelog reference to a list, with the first one being the URL to the online version, and the second being a literal string `''$src/doc/tour-changes.html''`, which is easy for humans to understand and will not be reached by tools as long as the first one is up.

Maybe `"\${src}/doc/tour-changes.html"` is a bit better fallback, no opinion here.

**Pull request merged**
